### ISSUE TITLE
I'm not able to install pulumi-esc-sdk on macOS

### ISSUE THREAD

**Issue by firemanxbr** [pulumi/esc-sdk#57]
### What happened?

when I added `pulumi-esc-sdk` in my `pyproject.toml` started getting the error below:

pulumi install --reinstall
Installing dependencies...

Resolved 34 packages in 13ms
Uninstalled 1 package in 3ms
Installed 1 package in 2ms
 - pulumi-esc-sdk==0.10.3
 + pulumi-esc-sdk==0.10.4
Finished installing dependencies

Downloading plugin command-1.0.1: 15.10 MiB / 15.10 MiB [===========] 100.00% 1s

[resource plugin command-1.0.1] installing
Downloading plugin pulumiservice-0.27.3: 14.34 MiB / 14.34 MiB [====] 100.00% 1s

[resource plugin pulumiservice-0.27.3] installing
Downloading plugin cloudflare-5.44.0: 18.89 MiB / 18.89 MiB [=======] 100.00% 2s

[resource plugin cloudflare-5.44.0] installing
Downloading plugin tls-5.0.9: 28.24 MiB / 28.24 MiB [===============] 100.00% 3s

[resource plugin tls-5.0.9] installing
Downloading plugin random-4.16.7: 27.92 MiB / 27.92 MiB [===========] 100.00% 3s

[resource plugin random-4.16.7] installing
Downloading plugin gcp-8.10.2: 37.61 MiB / 37.61 MiB [==============] 100.00% 3s

[resource plugin gcp-8.10.2] installing
Downloading plugin kubernetes-4.18.4: 63.88 MiB / 63.88 MiB [=======] 100.00% 4s

[resource plugin kubernetes-4.18.4] installing
Downloading plugin aws-6.64.0: 136.85 MiB / 136.85 MiB [============] 100.00% 5s

[resource plugin aws-6.64.0] installing
error: failed to download plugin: esc-sdk-0.10.4: failed to download plugin: esc-sdk-0.10.4: 403 HTTP error fetching plugin from https://get.pulumi.com/releases/plugins/pulumi-resource-esc-sdk-v0.10.4-darwin-arm64.tar.gz

### Example

checking the url -> https://get.pulumi.com/releases/plugins/pulumi-resource-esc-sdk-v0.10.4-darwin-arm64.tar.gz
I'm getting such error:
`This XML file does not appear to have any style information associated with it. The document tree is shown below.
<Error>
<Code>AccessDenied</Code>
<Message>Access Denied</Message>
<RequestId>DYAJV1W7NJQFNGX6</RequestId>
<HostId>zI8MPGXztOa6sTeVSYuitl4QzZWY0vf4npqB28emwISlqXmYf2CQKUiJZMvnyhh0P6y39rUdpkQ=</HostId>
</Error>`

### Output of `pulumi about`

pulumi about
CLI
Version      3.142.0
Go Version   go1.23.3
Go Compiler  gc

Plugins
KIND      NAME           VERSION
resource  aws            6.64.0
resource  cloudflare     5.44.0
resource  command        1.0.1
resource  esc-sdk        0.10.4
resource  gcp            8.10.2
resource  kubernetes     4.18.4
resource  pulumiservice  0.27.3
language  python         unknown
resource  random         4.16.7
resource  tls            5.0.9

Host
OS       darwin
Version  15.1.1
Arch     arm64

This project is written in python: executable='/Users/firemanxbr/Documents/work/iac/.venv/bin/python3' version='3.13.1'

Current Stack: nayaone/iac/n1-cloud-433112-hje74o-tools-green

Found no resources associated with nayaone/n1-cloud-433112-hje74o-tools-green

Found no pending operations associated with nayaone/n1-cloud-433112-hje74o-tools-green

Backend
Name           pulumi.com
URL            https://app.pulumi.com/marcelo-nayaone
User           marcelo-nayaone
Organizations  marcelo-nayaone, nayaone
Token type     personal

Dependencies:
NAME                  VERSION
annotated-types       0.7.0
arpeggio              2.0.2
attrs                 24.2.0
certifi               2024.8.30
charset-normalizer    3.4.0
debugpy               1.8.8
dill                  0.3.9
grpcio                1.66.2
idna                  3.10
nodeenv               1.9.1
parver                0.5
protobuf              4.25.5
pulumi                3.142.0
pulumi-aws            6.64.0
pulumi-cloudflare     5.44.0
pulumi-command        1.0.1
pulumi-esc-sdk        0.10.4
pulumi-gcp            8.10.2
pulumi-kubernetes     4.18.4
pulumi-pulumiservice  0.27.3
pulumi-random         4.16.7
pulumi-tls            5.0.9
pydantic              2.10.3
pydantic-core         2.27.1
pyright               1.1.390
python-dateutil       2.9.0.post0
pyyaml                6.0.2
requests              2.32.3
ruff                  0.8.2
semver                2.13.0
six                   1.16.0
typing-extensions     4.12.2
urllib3               2.0.7

Pulumi locates its logs in /var/folders/sj/0fh6n9750nl_075nfdwrpzh80000gn/T/ by default

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by seanyeh** [comment]
@firemanxbr I'm not able to reproduce this error. The error here looks the same to this previous [issue](https://github.com/pulumi/esc-sdk/issues/33), which should have been fixed since 0.10.2. What does your `requirements.txt` look like?

We recently published version 0.10.5, could you try that version and let me know if you're still running into the issue?

**Comment by firemanxbr** [comment]
@seanyeh I'm getting the same issue, looks like an issue in the pulumi repository. If you tried to access this URL: https://get.pulumi.com/releases/plugins/pulumi-resource-esc-sdk-v0.10.5-darwin-arm64.tar.gz you can see the results below:![Image](https://github.com/user-attachments/assets/3ec0991e-400e-434a-9de6-f518be1dd0fd)

output of `pulumi install`
`✗ pulumi install --reinstall
Installing dependencies...

Resolved 34 packages in 521ms
Prepared 1 package in 140ms
Uninstalled 1 package in 12ms
Installed 1 package in 3ms
 - pulumi-esc-sdk==0.10.4
 + pulumi-esc-sdk==0.10.5
Finished installing dependencies

Downloading plugin pulumiservice-0.27.3: 14.34 MiB / 14.34 MiB [====] 100.00% 0s

[resource plugin pulumiservice-0.27.3] installing
Downloading plugin command-1.0.1: 15.10 MiB / 15.10 MiB [===========] 100.00% 0s

[resource plugin command-1.0.1] installing
Downloading plugin cloudflare-5.44.0: 18.89 MiB / 18.89 MiB [=======] 100.00% 1s

[resource plugin cloudflare-5.44.0] installing
Downloading plugin tls-5.0.9: 28.24 MiB / 28.24 MiB [===============] 100.00% 1s

[resource plugin tls-5.0.9] installing
Downloading plugin random-4.16.7: 27.92 MiB / 27.92 MiB [===========] 100.00% 1s

[resource plugin random-4.16.7] installing
Downloading plugin gcp-8.10.2: 37.61 MiB / 37.61 MiB [==============] 100.00% 1s

[resource plugin gcp-8.10.2] installing
Downloading plugin kubernetes-4.18.4: 63.88 MiB / 63.88 MiB [=======] 100.00% 2s

[resource plugin kubernetes-4.18.4] installing
Downloading plugin aws-6.64.0: 136.85 MiB / 136.85 MiB [============] 100.00% 3s

[resource plugin aws-6.64.0] installing
error: failed to download plugin: esc-sdk-0.10.5: failed to download plugin: esc-sdk-0.10.5: 403 HTTP error fetching plugin from https://get.pulumi.com/releases/plugins/pulumi-resource-esc-sdk-v0.10.5-darwin-arm64.tar.gz`

**Comment by firemanxbr** [comment]
@seanyeh thanks for your quick response. I reviewed the issue I opened earlier [1]. The problem remains the same: I can't install esc-sdk using `pulumi install`, which seems very strange to me, because according to the documentation [2] and the article [3] these steps would be necessary:

```
pip install pulumi-esc-sdk
```

Import the sdk:
```
import pulumi_esc_sdk as esc
```
But how should I manage this library in my pulumi project? Because if I add it to requirements.txt or `pyproject.toml`, the `pulumi install` command will consume this dependency. How does the esc-sdk project expect users to use it? Outside of a Pulumi project? It doesn't seem viable to me, especially considering that the only other alternative would be to use the pulumiservice, which has no way of obtaining secrets stored in a Pulumi ESC Environment. I can use the `pulumi_pulumiservice.Environment.get` method to get the `yaml` data, but it doesn't open the Environment for me, I get a non-useful value like:

```
{"values": {"cloudflare": {"apiToken": {"fn::secret": {ciphertext: ZXNjeAAAAAAEAAAEAsjQlK5MMhrjrobct3WUZXXeQ3F9tJ43OFPZlRorvlRSvx9HtzXw2KS1fzud/TCN7HFslN/iwCpn59ROBhEIFKYFBgtFPy8kQ}},
```
This puts me back in a situation where changing the code from esc-sdk to multiservice keeps me locked out, and I have no access to the credentials I store in Pulumi ESC.

[1] - https://github.com/pulumi/esc-sdk/issues/33).
[2] - https://www.pulumi.com/docs/esc/development/languages-sdks/python/
[3] - https://www.pulumi.com/blog/esc-sdk-launch/

**Comment by firemanxbr** [comment]
My suggestion will be to add `open_and_read_environment()` function in `esc-sdk` to the `pulumiservice` to unblock me and possibly other users with the same use case.

What will solve two issues:

- Pulumiservice can be used to create and retrieve credentials from Pulumi ESC
- Pulumiservice does not need PULUMI_ACCESS_TOKEN as a dependency

**Comment by seanyeh** [comment]
@firemanxbr The error you're running into is showing that `pulumi` is trying to download esc-sdk as a pulumi plugin (which it is not) - that url https://get.pulumi.com/releases/plugins/pulumi-resource-esc-sdk-v0.10.5-darwin-arm64.tar.gz should not exist

You should be able to use the esc-sdk in a pulumi project. I'm looking into this and will get back to you as soon as I can.
In the meantime, are you able to provide a sample project that can reproduce the issue? I am not able to reproduce this with a new blank pulumi python program with `pulumi-esc-sdk` in my `requirements.txt`.

**Comment by firemanxbr** [comment]
@seanyeh for sure. I created a template project to allow you to test; feel free to clone -> https://github.com/firemanxbr/template-esc-sdk

**tl;dr;**
- `pulumi install`
- `pulumi preview`

```template-esc-sdk git:(main) pulumi install
Installing dependencies...

Using CPython 3.13.1 interpreter at: /opt/homebrew/opt/python@3.13/bin/python3.13
Creating virtual environment at: .venv
Resolved 19 packages in 42ms
Installed 18 packages in 30ms
 + annotated-types==0.7.0
 + debugpy==1.8.9
 + dill==0.3.9
 + grpcio==1.66.2
 + nodeenv==1.9.1
 + protobuf==4.25.5
 + pulumi==3.142.0
 + pulumi-esc-sdk==0.10.5
 + pydantic==2.10.3
 + pydantic-core==2.27.1
 + pyright==1.1.390
 + python-dateutil==2.9.0.post0
 + pyyaml==6.0.2
 + ruff==0.8.2
 + semver==2.13.0
 + six==1.17.0
 + typing-extensions==4.12.2
 + urllib3==2.0.7
Finished installing dependencies

error: failed to download plugin: esc-sdk-0.10.5: failed to download plugin: esc-sdk-0.10.5: 403 HTTP error fetching plugin from https://get.pulumi.com/releases/plugins/pulumi-resource-esc-sdk-v0.10.5-darwin-arm64.tar.gz```

**Comment by seanyeh** [comment]
@firemanxbr Thanks for the sample project, I was finally able to reproduce. I found out that it's due to a bug with the Pulumi CLI and `uv`. I've merged a quick fix for this to the Pulumi CLI, so this should be fixed in their next release. You can also build from source if you don't want to wait for the next release. Please let me know if you're still running into any issues

**Comment by firemanxbr** [comment]
@seanyeh thanks a lot. I went back to `pulumiservice` using `environmentVariables` it is a bit tricky to use environment variables to retrieve information from the Pulumi ESC environment, but it seems like a more straightforward way than using `pulumiservice` to create a new environment and using `esc-sdk` to retrieve secrets from an environment. With this approach, I managed to unblock myself. A small example below:

**stack 1** - creates a new environment:
```
env_json = pulumi.Output.json_dumps(
        {
            "values": {
                "cloudflare": {
                    "apiToken": {"fn::secret": api_token},
                    "account_id": data.get("account_id"),
                    "domain_name": data.get("domains")["name"],
                    "zone_id": data.get("domains")["zone_id"],
                },
                "pulumiConfig": {
                    "cloudflare:apiToken": "${cloudflare.apiToken}",
                },
                "environmentVariables": {
                    "CLOUDFLARE_API_TOKEN": "${cloudflare.apiToken}",
                    "CLOUDFLARE_ACCOUNT_ID": "${cloudflare.account_id}",
                    "CLOUDFLARE_DOMAIN_NAME": "${cloudflare.domain_name}",
                    "CLOUDFLARE_ZONE_ID": "${cloudflare.zone_id}",
                }
            },
        }
    )

    env_content = env_json.apply(lambda json_str: pulumi.StringAsset(json_str))

    pulumi_env = Environment(
        resource_name=tags.get("provider"),
        organization=org,
        name=tags.get("provider"),
        yaml=env_content,
        project=tags.get("provider"),
    )

    return pulumi_env.id
```

**stack 2** - consumes credentials created in **stack 1**
```
api_token = os.getenv("CLOUDFLARE_API_TOKEN")
    account_id = os.getenv("CLOUDFLARE_ACCOUNT_ID")

    if not api_token:
        raise ValueError(
            "CLOUDFLARE_API_TOKEN must be set using Pulumi ESC -> cloudflare/cloudflare!"
        )

    if not account_id:
        raise ValueError(
            "CLOUDFLARE_ACCOUNT_ID must be set using Pulumi ESC -> cloudflare/cloudflare!"
        )
```

Via code review, we can control the use of environments via stacks. If we could have RBAC for Pulumi, ESC like Pulumi Stacks would be amazing. We needed to keep developers out of ESC until we could have more granulation in the access control.
